scripts/vad_overlap_posterior.py

```diff
@@ -0,0 +1,268 @@
+   
+Usage:
+
+python vad_overlap_posterior.py --gen_overlap_seq --gen_seg_table --frame_folder=<FULL PATH OF YOU STORED FRAME LEVEL PREDICTION> --method='median' --overlap=0.875 --num_workers=20 --threshold=0.8 
+ 
+"""
+
+
+def gen_overlap_seq(frame_filepath, per_args):
+    """
+    Given a frame level prediction, generate predictions with overlapping input segments by using it
+    Args:
+        frame_filepath : frame prediction file to be processed.
+        per_args:
+            method : Median or mean smoothing filter.
+            overlap : Amounts of overlap.
+            seg_len : Length of window for generating the frame.
+            shift_len : Amount of shift of window for generating the frame.
+            out_dir : Output dir of generated prediction.
+    """
+
+    try:
+        method = per_args['method']
+        overlap = per_args['overlap']
+        seg_len = per_args['seg_len']
+        shift_len = per_args['shift_len']
+        out_dir = per_args['out_dir']
+
+        frame = np.loadtxt(frame_filepath)
+        name = os.path.basename(frame_filepath).split(".frame")[0] + "." + method
+        overlap_filepath = os.path.join(out_dir, name)
+
+        shift = int(shift_len / 0.01)  # number of units of shift
+        seg = int((seg_len / 0.01 + 1))  # number of units of each window/segment
+
+        jump_on_target = int(seg * (1 - overlap))  # jump on target generated sequence
+        jump_on_frame = int(jump_on_target / shift)  # jump on input frame sequence
+
+        if jump_on_frame <= 1:
+            raise Error("Not valid jump on frame sequence. Try different shift and overlap choices")
+
+        target_len = int(len(frame) * shift) + int(seg_len / 0.01)
+
+        if method == 'mean':
+            preds = np.zeros(target_len)
+            pred_count = np.zeros(target_len)
+
+            for i, og_pred in enumerate(frame):
+                if i % jump_on_frame != 0:
+                    continue
+                start = i * shift
+                end = start + seg
+                preds[start:end] = preds[start:end] + og_pred
+                pred_count[start:end] = pred_count[start:end] + 1
+
+            preds = preds / pred_count
+            last_non_zero_pred = preds[pred_count != 0][-1]
+            preds[pred_count == 0] = last_non_zero_pred
+
+            preds = preds[:-1]
+
+        elif method == 'median':
+            preds = [[] for _ in range(target_len)]
+            for i, og_pred in enumerate(frame):
+                if i % jump_on_frame != 0:
+                    continue
+                start = i * shift
+                end = start + seg
+                for i in range(start, end):
+                    preds[i].append(og_pred)
+
+            preds = np.array([np.median(l) for l in preds])
+
+            nan_idx = np.isnan(preds)
+            last_non_nan_pred = preds[~nan_idx][-1]
+            preds[nan_idx] = last_non_nan_pred
+            preds = preds[:-1]
+
```

Comment:
Not convinced that median filtering is useful for VAD. Even if majority of the frames are silence/non-speech, you would want to classify the segment as speech so that you don't drop any speech segments, so there is a risk of losing most Beginning Of Sentence (BOS) and EOS segments with median filtering. Do you see how this is useful? Or are there any other implementations using this kind of filtering for VAD?